### PR TITLE
Add `PauliString`

### PR DESCRIPTION
**Pull request by rmlarose**
Description
-----------

Part of #623. This is a thin wrapper for `cirq.PauliString` with appropriate I/O for us.

Checklist
-----------

Check off the following once complete (or if not applicable). The PR will be reviewed once this
checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [ ] I updated the [changelog](https://github.com/unitaryfund/mitiq/blob/master/CHANGELOG.md) including author and PR number (@username, gh-xxx)

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

Tips
----

- If the validation check fails:
  1. Run `make check-style` (from the root directory
  of the repository) and fix any [flake8](http://flake8.pycqa.org) errors.
  2. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html)
  autoformatter.
- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

File: mitiq/observable/pauli.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                )
+            measured = (
+                circuit + pauli.to_z_basis_ops() + cirq.measure(*pauli.qubits)
+            )
+
+            # Transform circuit back to original qubits.
+            reverse_qubit_map = dict(zip(qubit_map.values(), qubit_map.keys()))
+            return measured.transform_qubits(lambda q: reverse_qubit_map[q])
+
+        return _measure_in(circuit, self._pauli)
```
Comment (andreamari) [inline mitiq/observable/pauli.py:102]: This function `measure_in` is nice but I think it measures the Pauli operator without considering the coefficient.  E.g. it acts in the same way when measuring XZ or 7*XZ. 
I am not sure what is the best solution, e.g.:
- we add a note in the docstring of `measure_in` saying that the coefficient is neglected.
- we could force `PauliString` to always have coefficient 1 and store the coefficients in a different place/object.

**Review by andreamari — APPROVED** [review]
Hi @rmlarose, sorry for the long delay, I forgot about it.
LGTM apart from the comment about `measure_in`.

A general comment: do we want the user to use this class?
If PauliString and other similar future objects will only be used internally by Mitiq, maybe we don't need to support circuit conversions.

File: mitiq/observable/pauli.py (hunk 1 above)
Comment (rmlarose) [reply mitiq/observable/pauli.py:102]: The coefficient shouldn't go with the circuit but rather the results from executing the circuit. In other words, the circuits you would run to measure XZ or 7*XZ are the same, so they should be the same here. Perhaps there is a better name than `measure_in`? Or I can add a note to the docstring.

**Comment by rmlarose** [comment]
> A general comment: do we want the user to use this class?
> If PauliString and other similar future objects will only be used internally by Mitiq, maybe we don't need to support circuit conversions.

I think the workflow should be identical for both `Observable` and `PauliString` and so this class should be public.

File: mitiq/observable/pauli.py (hunk 1 above)
Comment (rmlarose) [reply mitiq/observable/pauli.py:102]: Verbose, but what this method really does is `add_single_qubit_basis_rotations_to_measure_in(circuit: QPROGRAM) -> QPROGRAM`. Or we could steal the Cirq name of `to_z_basis_ops`. Thoughts?

Comment (andreamari) [reply mitiq/observable/pauli.py:102]: With `to_z_basis_ops` one would expect only the final rotation, not the full circuit. So `measure_in` is fine.

**Pull request merged**